featurebyte/service/offline_store_feature_table.py

```diff
@@ -65,23 +139,73 @@ async def create_document(
         -------
         OfflineStoreFeatureTableModel
         """
-        if data.entity_lookup_info is None:
-            # check if name already exists
-            data.name = data.get_name()
-            query_filter = {"name": data.name}
-            query_result = await self.list_documents_as_dict(query_filter=query_filter, page_size=1)
+        with self.get_feature_cluster_storage_lock(
+            data.id, timeout=OFFLINE_STORE_FEATURE_TABLE_REDIS_LOCK_TIMEOUT
+        ):
+            return await self._create_document(data)
 
-            count = query_result["total"]
-            data.name_suffix = None
-            if count:
-                # if name already exists, append a suffix
-                data.name_suffix = str(count)
-                data.name = data.get_name()
-
-        output = await super().create_document(data)
-        assert output.catalog_id == data.catalog_id
+    async def _update_offline_feature_table(
+        self,
+        document_id: ObjectId,
+        data: OfflineStoreFeatureTableUpdate,
+        exclude_none: bool = True,
+        skip_block_modification_check: bool = False,
+    ) -> OfflineStoreFeatureTableModel:
+        original_doc = await self.get_document(
+            document_id=document_id, populate_remote_attributes=False
+        )
+        if isinstance(data, FeaturesUpdate):
+            assert (
+                data.feature_cluster_path is None
+            ), "feature_cluster_path should not be set in update"
+            if original_doc.feature_cluster_path:
+                # attempt to remove the old feature cluster
+                await self.storage.try_delete_if_exists(Path(original_doc.feature_cluster_path))
+
+            table = OfflineStoreFeatureTableModel(
+                **{**original_doc.dict(by_alias=True), **data.dict(by_alias=True)}
+            )
+            table = await self._move_feature_cluster_to_storage(table)
+            data.feature_cluster = None
+            data.feature_cluster_path = table.feature_cluster_path
+
+        output = await super().update_document(
+            document_id,
+            data,
+            exclude_none=exclude_none,
+            document=original_doc,
+            skip_block_modification_check=skip_block_modification_check,
+        )
```

Comment:
If the mongo update fail, would the already moved remote feature cluster be in a potentially inconsistent state?

Reply:
Yes, this is possible. I think we just need to make sure second run using the same update should not fail (so that we can add `retry` decorator).